horovod/common/operations.cc

```diff
@@ -694,6 +694,36 @@ bool horovod_is_initialized() {
   return horovod_global.initialization_done;
 }
 
+bool horovod_start_timeline(const char* file_name, bool mark_cycles) {
+  if (!horovod_global.initialization_done) {
+    return false;
+  }
+
+  bool is_coordinator = horovod_global.controller->IsCoordinator();
+  if (is_coordinator) {
+    horovod_global.timeline.Initialize(file_name, horovod_global.controller->GetSize());
+  }
+
+  horovod_global.mark_cycles_in_timeline = mark_cycles;
+  horovod_global.controller->SetTimelineEnabled(true);
+  return true;
+}
+
+bool horovod_stop_timeline() {
+  if (!horovod_global.initialization_done) {
+    return false;
+  }
+
+  horovod_global.controller->SetTimelineEnabled(false);
```

Comment:
i guess you need to make this function(SetTimelineEnabled) thread safe. As there would be 2 different processes accessing this variable. One process trying to write the value while at the same time other process can be reading the timeline_enabled_

https://github.com/horovod/horovod/blob/e4554de96100f5a0e8686cd41cf99a6fe8a71e62/horovod/common/controller.cc#L661 and L 670 in same file

Reply:
I also think that you should start timeline when horovod is inited with SetTimelineEnabled is false if env is not set. 

This way you would just need to protect the race for timeline_Enabled_ which can be concurrently accessed.